### PR TITLE
K8SPSMDB-513 feature: Add support of Microsoft Azure Blob storage for the backups

### PR DESCRIPTION
**Pull request by pooknull**
[K8SPSMDB-513](https://jira.percona.com/browse/K8SPSMDB-513) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

https://jira.percona.com/browse/K8SPSMDB-513

### CODE REVIEW

**Comment by pooknull** [comment]
Waiting for #799 to be merged

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | failed |
| storage | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| limits | passed |
| demand-backup | failed |
| scheduled-backup | failed |
| scaling | passed |
| demand-backup-sharded | failed |
| security-context | passed |
| upgrade | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| rs-shard-migration | passed |
| upgrade-sharded | passed |
| pitr | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| pitr-sharded | failed |
| users | failed |
| data-sharded | passed |
| non-voting | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/800/commits/73bb46f0e5b949cb03b7caca2e93ba259b8a2863
image: `perconalab/percona-server-mongodb-operator:PR-800-73bb46f0`

**Comment by defbin** [comment]
> Waiting for #799 to be merged

it should be fixed by https://github.com/percona/percona-server-mongodb-operator/pull/801

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| limits | passed |
| monitoring-2-0 | passed |
| demand-backup | failed |
| scaling | passed |
| scheduled-backup | failed |
| arbiter | passed |
| demand-backup-sharded | failed |
| security-context | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| upgrade | passed |
| rs-shard-migration | passed |
| service-per-pod | passed |
| upgrade-sharded | passed |
| liveness | passed |
| smart-update | passed |
| pitr | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |
| non-voting | passed |
| pitr-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/800/commits/0beec77d18e3082b0a5155192ae7654e4b54229b
image: `perconalab/percona-server-mongodb-operator:PR-800-0beec77d`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| demand-backup | failed |
| limits | passed |
| monitoring-2-0 | passed |
| scheduled-backup | failed |
| scaling | passed |
| demand-backup-sharded | failed |
| arbiter | passed |
| security-context | passed |
| upgrade | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| rs-shard-migration | passed |
| service-per-pod | passed |
| upgrade-sharded | passed |
| liveness | passed |
| pitr | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |
| non-voting | passed |
| pitr-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/800/commits/e3250b6327a1f317c294f92c6e2447eb0c497d02
image: `perconalab/percona-server-mongodb-operator:PR-800-e3250b63`

File: e2e-tests/scheduled-backup/run [hunk 1]
```diff
@@ -7,13 +7,13 @@ test_dir=$(realpath $(dirname $0))
 . ${test_dir}/../functions
```
Comment (delgod) [inline e2e-tests/scheduled-backup/run:7]: please roll back the file permissions.

File: pkg/apis/psmdb/v1/psmdb_types.go [hunk 2]
```diff
@@ -533,16 +533,25 @@ type BackupStorageS3Spec struct {
 	CredentialsSecret string `json:"credentialsSecret"`
 }
 
+type BackupStorageAzureSpec struct {
+	Account           string `json:"account,omitempty"`
```
Comment (delgod) [inline pkg/apis/psmdb/v1/psmdb_types.go:537]: maybe we should take account id from secret. please ask @spron-in

**Review by delgod — CHANGES_REQUESTED** [review]

File: e2e-tests/demand-backup-sharded/conf/some-name-rs0.yml [hunk 3]
```diff
@@ -34,6 +34,14 @@ spec:
           bucket: operator-testing
           prefix: psmdb
           endpointUrl: https://storage.googleapis.com
+      azure-blob:
+        type: azure
+        azure:
+          account: pxcperator
```
Comment (hors) [inline e2e-tests/demand-backup-sharded/conf/some-name-rs0.yml:40]:
```suggestion
          account: k8soperators
```

File: e2e-tests/demand-backup-sharded/conf/some-name-rs0.yml [hunk 4]
```diff
@@ -34,6 +34,14 @@ spec:
...
+          container: operato-test-storage
```
Comment (hors) [inline e2e-tests/demand-backup-sharded/conf/some-name-rs0.yml:41]:
```suggestion
          container: operator-testing
```

File: e2e-tests/demand-backup/conf/some-name-rs0.yml [hunk 5]
```diff
@@ -34,6 +34,14 @@ spec:
           bucket: operator-testing
           prefix: psmdb
           endpointUrl: https://storage.googleapis.com
+      azure-blob:
+        type: azure
+        azure:
+          account: pxcperator
```
Comment (hors) [inline e2e-tests/demand-backup/conf/some-name-rs0.yml:40]:
```suggestion
          account: k8soperators
```

File: e2e-tests/demand-backup/conf/some-name-rs0.yml [hunk 6]
```diff
@@ -34,6 +34,14 @@ spec:
...
+          container: operato-test-storage
```
Comment (hors) [inline e2e-tests/demand-backup/conf/some-name-rs0.yml:41]:
```suggestion
          container: operator-testing
```

**Review by hors — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| limits | passed |
| demand-backup | failed |
| monitoring-2-0 | passed |
| scaling | passed |
| arbiter | passed |
| security-context | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| rs-shard-migration | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| scheduled-backup | failed |
| demand-backup-sharded | failed |
| users | passed |
| data-sharded | passed |
| non-voting | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| pitr | passed |
| pitr-sharded | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/800/commits/723172a8f03c960e51de797262e3ef873630298b
image: `perconalab/percona-server-mongodb-operator:PR-800-723172a8`

File: pkg/apis/psmdb/v1/psmdb_types.go (hunk 2 above)
Comment (hors) [reply pkg/apis/psmdb/v1/psmdb_types.go:537]: Please use the following keys for secret:
```
 AZURE_STORAGE_ACCOUNT_NAME:
  AZURE_STORAGE_ACCOUNT_KEY: 
```

File: e2e-tests/demand-backup-sharded/conf/some-name-rs0.yml (hunk 3 above)
Comment (pooknull) [reply e2e-tests/demand-backup-sharded/conf/some-name-rs0.yml]: Resolved in e23e34aa6fe192bf81d1378f5c0e3fb76ec60d85

File: e2e-tests/scheduled-backup/run (hunk 1 above)
Comment (pooknull) [reply e2e-tests/scheduled-backup/run:7]: Resolved in 9621b66f1ebeeb2c13c2c436d2b89f49332f22a0

File: pkg/apis/psmdb/v1/psmdb_types.go (hunk 2 above)
Comment (pooknull) [reply pkg/apis/psmdb/v1/psmdb_types.go]: Resolved in 054f92f163b58ce537da894e76dbdeba575b8e71

**Review by delgod — DISMISSED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| monitoring-2-0 | passed |
| limits | passed |
| demand-backup | failed |
| scaling | passed |
| arbiter | passed |
| security-context | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| scheduled-backup | failed |
| users | passed |
| data-sharded | passed |
| non-voting | passed |
| demand-backup-sharded | failed |
| upgrade | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/800/commits/054f92f163b58ce537da894e76dbdeba575b8e71
image: `perconalab/percona-server-mongodb-operator:PR-800-054f92f1`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| monitoring-2-0 | passed |
| limits | passed |
| demand-backup | failed |
| scaling | passed |
| arbiter | passed |
| security-context | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| rs-shard-migration | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| scheduled-backup | failed |
| users | passed |
| data-sharded | passed |
| demand-backup-sharded | failed |
| non-voting | passed |
| upgrade | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/800/commits/7bae7fe6d75b7419cababb8b6f3cb48c094ed3e3
image: `perconalab/percona-server-mongodb-operator:PR-800-7bae7fe6`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/800/commits/7bae7fe6d75b7419cababb8b6f3cb48c094ed3e3
image: `perconalab/percona-server-mongodb-operator:PR-800-7bae7fe6`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| monitoring-2-0 | failed |
| demand-backup | failed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| rs-shard-migration | passed |
| arbiter | passed |
| service-per-pod | passed |
| scheduled-backup | failed |
| demand-backup-sharded | failed |
| upgrade | failed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/800/commits/a742472715d41cdea7f94a54375f1ded655f9265
image: `perconalab/percona-server-mongodb-operator:PR-800-a7424727`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| rs-shard-migration | passed |
| one-pod | passed |
| storage | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| upgrade-consistency | passed |
| demand-backup | failed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| scheduled-backup | passed |
| demand-backup-sharded | failed |
| data-sharded | passed |
| non-voting | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| pitr | passed |
| pitr-sharded | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/800/commits/a742472715d41cdea7f94a54375f1ded655f9265
image: `perconalab/percona-server-mongodb-operator:PR-800-a7424727`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| rs-shard-migration | passed |
| one-pod | passed |
| monitoring-2-0 | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| storage | passed |
| data-sharded | passed |
| self-healing | passed |
| non-voting | passed |
| self-healing-chaos | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | failed |
| upgrade | passed |
| upgrade-sharded | passed |
| pitr | passed |
| pitr-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/800/commits/a742472715d41cdea7f94a54375f1ded655f9265
image: `perconalab/percona-server-mongodb-operator:PR-800-a7424727`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| limits | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| scaling | passed |
| arbiter | passed |
| security-context | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| scheduled-backup | passed |
| rs-shard-migration | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| data-sharded | passed |
| pitr | passed |
| non-voting | passed |
| pitr-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/800/commits/be96c15336d96d6f2163c4a4a0e3b4f278e16c35
image: `perconalab/percona-server-mongodb-operator:PR-800-be96c153`

**Review by hors — APPROVED** [review]

File: e2e-tests/functions [hunk 7]
```diff
@@ -242,6 +242,7 @@ wait_restore() {
 	echo
 	set -o xtrace
 
+  sleep 20
```
[the remark this answers is not in the document]
Comment (delgod) [reply e2e-tests/functions:245]: valid comment

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | failed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| monitoring-2-0 | passed |
| limits | passed |
| demand-backup | passed |
| scaling | passed |
| arbiter | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| security-context | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| liveness | passed |
| demand-backup-sharded | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| data-sharded | passed |
| pitr | passed |
| non-voting | passed |
| pitr-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/800/commits/3a5f437cbcf626fafdfe2e6d088c43507a54ca0b
image: `perconalab/percona-server-mongodb-operator:PR-800-3a5f437c`

**Review by defbin — APPROVED** [review]
lgtm

**Comment by defbin** [comment]
@pooknull I see many indention changes. could you please make sure you did it correctly?

**Comment by defbin** [comment]
run this https://github.com/percona/percona-server-mongodb-operator/blob/main/.github/workflows/reviewdog.yml#L34-L35 for exactly changed in the PR shell scripts. like
```
$(go env GOPATH)/bin/shfmt -bn -ci -s -w e2e-tests/functions
```

**Pull request merged**